### PR TITLE
Feature/websockets

### PR DESCRIPTION
**Pull request by hedefalk**
Replaced the old swank envelopes with websocket.

### CODE REVIEW

**Comment by hedefalk** [comment]
@d6y Got time for peer review?

**Comment by d6y** [comment]
Looking...

**Comment by d6y** [comment]
General comment, probably not specific to this PR.  I've not built this project before, and when I did there are some warnings:

```
ensime-node (hedefalk-feature/websockets *)$ npm install
npm WARN deprecated lodash@0.9.2: Grunt needs your help! See https://github.com/gruntjs/grunt/issues/1403.
npm WARN deprecated graceful-fs@1.2.3: graceful-fs version 3 and before will fail on newer node releases. Please update to graceful-fs@^4.0.0 as soon as possible.
npm WARN deprecated lodash@1.0.2: lodash@<3.0.0 is no longer maintained. Upgrade to lodash@^4.0.0.
npm WARN deprecated lodash@2.4.1: lodash@<3.0.0 is no longer maintained. Upgrade to lodash@^4.0.0.
npm WARN prefer global jasmine-node@1.10.2 should be installed with -g
```

BTW, I notice you've updated lodash in package.js so no idea what's causing the above. Another dependency maybe?

File: src/client.coffee [hunk 1]
```diff
@@ -1,163 +1,158 @@
...
 
-module.exports =
-class Client
-  constructor: (port, @httpPort, generalMsgHandler, @serverPid = undefined) ->
-    @ensimeMessageCounter = 1
-    @callbackMap = {}
-
-    @parser = new Swank.SwankParser( (env) =>
-      log.trace("incoming: #{env}")
+module.exports = createClient = (port, httpPort, generalMsgHandler, serverPid = undefined) ->
```
Comment (d6y) [inline src/client.coffee]: `port` can be removed, maybe? Here and where it is read in _ensime-client-startup.coffee_ ?

File: src/client.coffee [hunk 2]
```diff
@@ -1,163 +1,158 @@
...
+    
+    callbackMap = {}
+
+    ensimeMessageCounter = 1
+      
+    websocket = new WebSocket("ws://localhost:" + httpPort + "/jerky")
+
+    websocket.on "open", ->
+      log.trace "connecting websocket…"
+      resolve(publicApi()) # FIXME
```
Comment (d6y) [inline src/client.coffee]: Can this `FIXME` go, or be moved into a ticket?

File: src/client.coffee [hunk 3]
```diff
@@ -1,163 +1,158 @@
...
-
-  symbolByName: (qualifiedName, callback) ->
-    msg =
-      typehint: 'SymbolByNameReq'
-      typeFullName: qualifiedName
-    @post(msg, callback)
     
+    postString = (msg, callback) ->
+      msg = """{"req": #{msg}, "callId": #{ensimeMessageCounter}}"""
+      callbackMap[ensimeMessageCounter++] = callback
```
Comment (d6y) [inline src/client.coffee:71]: I'm worried about the `++` increment.  If we do one call  a second we're gong to run into [MAX_SAFE_INTEGER](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/MAX_SAFE_INTEGER) in just 285 million years.

Oh, come on. That was a good joke.

File: package.json [hunk 4]
```diff
@@ -7,7 +7,7 @@
     "prepublish": "grunt prepublish",
     "test": "grunt test",
     "it": "grunt it",
-    "watch": "grunt ts:dev"
+    "watch": "grunt watch"
   },
   "repository": "https://github.com/hedefalk/ensime-node",
```
Comment (d6y) [inline package.json]: This should now be _https://github.com/ensime/ensime-node_

File: src/client.coffee [hunk 5]
```diff
@@ -1,163 +1,158 @@
 net = require('net')
-Swank = require './lisp/swank-protocol'
 fs = require 'fs-extra'
 path = require 'path'
 log = require('loglevel').getLogger('ensime.client')
 temp = require 'temp'
+WebSocket = require("ws")
 
+temp.track()
```
Comment (d6y) [inline src/client.coffee:8]: neat! that's how you clear up files. I didn't know that.

**Comment by d6y** [comment]
@hedefalk I've looked, and made some trivial comments, but I've not run.

I was following the instructions on ensime.org and failed at the `apm rebuild` (Atom special sauce stage):

I guess I have the wrong version of something...

> node-pre-gyp ERR! stack Error: Unsupported target version: 0.34.5

blah blah

> Failed at the fsevents@1.0.9 install script 'node-pre-gyp install --fallback-to-build'.
> npm ERR! This is most likely a problem with the fsevents package,

```
ensime-node (hedefalk-feature/websockets *)$ apm rebuild
Rebuilding modules npm run watch
✗
npm WARN prefer global coffee-script@1.10.0 should be installed with -g
npm WARN prefer global jasmine-node@1.10.2 should be installed with -g
npm WARN prefer global typings@0.7.9 should be installed with -g
./Release/.deps/Release/obj.target/fse/fsevents.o.d.raw { dev: 16777220,
  mode: 33188,
  nlink: 1,
  uid: 502,
  gid: 20,
  rdev: 0,
  blksize: 4096,
  ino: 119733406,
  size: 1594,
  blocks: 8,
  atime: Wed Mar 23 2016 22:11:22 GMT+0000 (GMT),
  mtime: Wed Mar 23 2016 22:11:22 GMT+0000 (GMT),
  ctime: Wed Mar 23 2016 22:11:22 GMT+0000 (GMT) }
./Release/.deps/Release/obj.target/fse/fsevents.o.d.raw { dev: 16777220,
  mode: 33188,
  nlink: 1,
  uid: 502,
  gid: 20,
  rdev: 0,
  blksize: 4096,
  ino: 119733430,
  size: 1594,
  blocks: 8,
  atime: Wed Mar 23 2016 22:11:24 GMT+0000 (GMT),
  mtime: Wed Mar 23 2016 22:11:24 GMT+0000 (GMT),
  ctime: Wed Mar 23 2016 22:11:24 GMT+0000 (GMT) }
node-pre-gyp ERR! install error
node-pre-gyp ERR! stack Error: Unsupported target version: 0.34.5
node-pre-gyp ERR! stack     at get_runtime_abi (/Users/richard/Developer/third-party/ensime-node/node_modules/fsevents/node_modules/node-pre-gyp/lib/util/versioning.js:156:23)
node-pre-gyp ERR! stack     at Object.module.exports.evaluate (/Users/richard/Developer/third-party/ensime-node/node_modules/fsevents/node_modules/node-pre-gyp/lib/util/versioning.js:265:19)
node-pre-gyp ERR! stack     at install (/Users/richard/Developer/third-party/ensime-node/node_modules/fsevents/node_modules/node-pre-gyp/lib/install.js:138:31)
node-pre-gyp ERR! stack     at Object.self.commands.(anonymous function) [as install] (/Users/richard/Developer/third-party/ensime-node/node_modules/fsevents/node_modules/node-pre-gyp/lib/node-pre-gyp.js:48:37)
node-pre-gyp ERR! stack     at run (/Users/richard/Developer/third-party/ensime-node/node_modules/fsevents/node_modules/node-pre-gyp/bin/node-pre-gyp:79:30)
node-pre-gyp ERR! stack     at Object.<anonymous> (/Users/richard/Developer/third-party/ensime-node/node_modules/fsevents/node_modules/node-pre-gyp/bin/node-pre-gyp:131:1)
node-pre-gyp ERR! stack     at Module._compile (module.js:456:26)
node-pre-gyp ERR! stack     at Object.Module._extensions..js (module.js:474:10)
node-pre-gyp ERR! stack     at Module.load (module.js:356:32)
node-pre-gyp ERR! stack     at Function.Module._load (module.js:312:12)
node-pre-gyp ERR! System Darwin 15.3.0
node-pre-gyp ERR! command "node" "/Users/richard/Developer/third-party/ensime-node/node_modules/fsevents/node_modules/.bin/node-pre-gyp" "install" "--fallback-to-build"
node-pre-gyp ERR! cwd /Users/richard/Developer/third-party/ensime-node/node_modules/fsevents
node-pre-gyp ERR! node -v v0.10.40
node-pre-gyp ERR! node-pre-gyp -v v0.6.24
node-pre-gyp ERR! not ok

npm ERR! Darwin 15.3.0
npm ERR! argv "/Applications/Atom.app/Contents/Resources/app/apm/bin/node" "/Applications/Atom.app/Contents/Resources/app/apm/node_modules/npm/bin/npm-cli.js" "--globalconfig" "/Users/richard/.atom/.apm/.apmrc" "--userconfig" "/Users/richard/.atom/.apmrc" "rebuild" "--target=0.34.5" "--arch=x64"
npm ERR! node v0.10.40
npm ERR! npm  v2.13.3
npm ERR! code ELIFECYCLE
npm ERR! fsevents@1.0.9 install: `node-pre-gyp install --fallback-to-build`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the fsevents@1.0.9 install script 'node-pre-gyp install --fallback-to-build'.
npm ERR! This is most likely a problem with the fsevents package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-pre-gyp install --fallback-to-build
npm ERR! You can get their info via:
npm ERR!     npm owner ls fsevents
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/richard/Developer/third-party/ensime-node/npm-debug.log

ensime-node (hedefalk-feature/websockets *)$
```

Complete log at: https://gist.github.com/d6y/cb56eee5e6e0131f4541

Maybe catch up with you tomorrow, if you're around (Thu 24 - then I'm away for a week)

File: src/client.coffee (hunk 1 above)
Comment (hedefalk) [inline src/client.coffee]: Ooh, yes! Port can go! Great!

File: src/client.coffee (hunk 2 above)
Comment (hedefalk) [inline src/client.coffee]: Yes, I think that was a "wip-comment" that's already fixed. I'll kill it.

File: src/client.coffee (hunk 3 above)
Comment (hedefalk) [inline src/client.coffee:71]: Hehe!

I actually really felt this was a dirty hack on an instinct level. There's some low instinct that I probably shouldn't make my own autoincrement id's, like in a sql db or whatever. Feels lika a lost opportunity too google and use a lib or something :)

File: package.json (hunk 4 above)
Comment (hedefalk) [inline package.json]: Yup!

**Comment by hedefalk** [comment]
Yeah, the apm rebuild fails because it's using an ancient version of npm. I tried to ask here: https://discuss.atom.io/t/apm-using-ancient-versions-of-node-npm-and-i-cant-seem-to-fix-it/26400

However, everything still seems to work so I haven't bothered. I actually haven't run apm rebuild for quite a bit. Can't seem to understand exactly what it does…

Related thing:
http://stackoverflow.com/questions/35798211/depending-on-lots-of-npm-from-vscode-module

**Comment by hedefalk** [comment]
The deprecated lodash is most certainly a transitive dependency. I'm still not sure how npm does that, but basically it doesn't necessarily flatten deps like mvn/ivy since you have "built in osgi" so to speak (just nested dirs under node_modules). But it CAN flatten if you want I think. I think you can lock a dependency in main modules package.json. I need to look into that stuff. Any day now… :)

**Pull request merged**